tests/trace/test_server_file_storage.py

```diff
@@ -0,0 +1,231 @@
+                    "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+                    "token_uri": "https://oauth2.googleapis.com/token",
+                    "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+                    "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/test@test-project.iam.gserviceaccount.com"
+                }""",
+                "WF_FILE_STORAGE_URI": f"gcs://{TEST_BUCKET}",
+            },
+        ):
+            yield
+
+    @pytest.mark.usefixtures("gcp_storage_env", "mock_gcp_credentials")
+    def test_gcp_storage(self, run_storage_test, gcs):
+        """Test file storage using Google Cloud Storage."""
+        res = run_storage_test()
+
+        # Verify the object exists in GCS
+        bucket = gcs.bucket(TEST_BUCKET)
+        blob = bucket.blob(res.digest)
+        assert blob.download_as_bytes() == TEST_CONTENT
+
+
+class TestAzureStorage:
+    """Tests for Azure Blob Storage implementation."""
+
+    @pytest.fixture
+    def azure_blob(self):
+        """Azure Blob Storage using Azurite emulator."""
+        conn_str = (
+            f"DefaultEndpointsProtocol=http;"
+            f"AccountName={AZURITE_ACCOUNT};"
+            f"AccountKey={AZURITE_KEY};"
+            f"BlobEndpoint={AZURITE_URL};"
+        )
+
+        # Create the Azurite client
+        azurite_client = BlobServiceClient.from_connection_string(conn_str)
+
+        # Mock the client creation to always return our Azurite client
+        with mock.patch("azure.storage.blob.BlobServiceClient") as mock_client:
+            mock_client.from_connection_string.return_value = azurite_client
+
+            # Create test container
+            try:
+                azurite_client.create_container(TEST_BUCKET)
+            except Exception:
+                # Container might already exist
+                pass
+
+            yield azurite_client
+
+    @pytest.fixture
+    def azure_storage_env(self):
+        """Setup Azure storage environment."""
+        with mock.patch.dict(
+            os.environ,
+            {
+                "WF_FILE_STORAGE_BUCKET_AZURE_CREDENTIAL_B64": AZURITE_B64_KEY,
+                "WF_FILE_STORAGE_BUCKET_AZURE_ACCOUNT_URL": AZURITE_URL,
+                "WF_FILE_STORAGE_URI": f"az://{AZURITE_ACCOUNT}/{TEST_BUCKET}",
+            },
+        ):
+            yield
+
+    @pytest.mark.usefixtures("azure_storage_env")
+    def test_azure_storage(self, run_storage_test, azure_blob):
+        """Test file storage using Azure Blob Storage."""
+        res = run_storage_test()
+
+        # Verify the object exists in Azure
+        container_client = azure_blob.get_container_client(TEST_BUCKET)
+        # Hard-coding project for ease. If we change how CI generates projects, this is ok to change
+        project = "c2hhd24vdGVzdC1wcm9qZWN0"
+        blob_client = container_client.get_blob_client(
+            f"weave/projects/{project}/files/{res.digest}"
```

Comment:
nit: this should use `key_for_project_digest`

Reply:
I prefer to not use the internal method for this as it would then propagate error of implementation into the test. For example, if i spelt "projects" incorrectly in the helper, then the test would pass